File: packages/core_dart/lib/src/tag_styles.dart
/// `tag_styles.dart` — KRO-30. `TagStyle` + validador puro contra el catálogo
/// de efectos visuales. Espejo de `tag-styles.ts` (+ el tipo `TagStyle` de
/// `types.ts`).
///
/// Pure: sin side effects, sin I/O. Mismas entradas → mismas issues.
library;

import 'visual_effects.dart';

/// Mapeo valor-de-tag → efecto visual. Vive en `albumSchema.tagStyles`. Una
/// carta cuya tag tenga este `value` dispara el `effect` (con `config`).
class TagStyle {
  /// Valor EXACTO de la tag que dispara el efecto (ej. "Holográfica").
  final String value;

  /// ID del efecto del catálogo (ej. "holographic_effect").
  final String effect;

  /// Config opcional del efecto. Keys + espacio de valores los define el efecto
  /// (valores `String | num`).
  final Map<String, Object>? config;

  const TagStyle({required this.value, required this.effect, this.config});

  factory TagStyle.fromJson(Map<String, dynamic> json) => TagStyle(
        value: (json['value'] as String?) ?? '',
        effect: (json['effect'] as String?) ?? '',
        config: (json['config'] as Map?)
            ?.map((k, v) => MapEntry(k.toString(), v as Object)),
      );
}

/// Un issue del validador de tag styles. Espejo de `TagStyleValidationIssue`.
class TagStyleValidationIssue {
  /// Índice del TagStyle en el array (-1 si se validó uno suelto).
  final int index;
  final String path;
  final String message;

  /// 'error' = no aplicable; 'warn' = aplicable con fallback.
  final String level;

  const TagStyleValidationIssue({
    required this.index,
    required this.path,
    required this.message,
    required this.level,
  });

  @override
  String toString() => '[$level] $path: $message';
}

/// Resultado de [validateTagStyles]. Espejo de `TagStyleValidationResult`.
class TagStyleValidationResult {
  final bool valid;
  final List<TagStyleValidationIssue> issues;
  const TagStyleValidationResult({required this.valid, required this.issues});
}

void _collectTagStyleIssues(
    TagStyle ts, int index, String prefix, List<TagStyleValidationIssue> issues) {
  // 1. value no vacío.
  if (ts.value.trim().isEmpty) {
    issues.add(TagStyleValidationIssue(
        index: index,
        path: '$prefix.value',
        level: 'error',
        message: 'el valor de la tag no puede estar vacío'));
  }

  // 2. effect existe en el catálogo.
  final effect = getVisualEffect(ts.effect);
  if (effect == null) {
    issues.add(TagStyleValidationIssue(
        index: index,
        path: '$prefix.effect',
        level: 'error',
        message:
            'el efecto "${ts.effect}" no existe en el catálogo de efectos visuales'));
    return; // sin definición no podemos validar config.
  }

  final params = effect.config;
  final paramByKey = {for (final p in params) p.key: p};
  final config = ts.config ?? const <String, Object>{};

  // 3. cada key de config pertenece al efecto.
  config.forEach((key, value) {
    final param = paramByKey[key];
    if (param == null) {
      issues.add(TagStyleValidationIssue(
          index: index,
          path: '$prefix.config.$key',
          level: 'error',
          message: '"$key" no es una opción de config del efecto "${ts.effect}"'));
      return;
    }
    _validateConfigValue(param, value, index, '$prefix.config.$key', issues);
  });

  // 4. params requeridos presentes.
  for (final param in params) {
    if (param.optional == false && !config.containsKey(param.key)) {
      issues.add(TagStyleValidationIssue(
          index: index,
          path: '$prefix.config.${param.key}',
          level: 'error',
          message: 'el efecto "${ts.effect}" requiere config "${param.key}"'));
    }
  }
}

void _validateConfigValue(VisualEffectConfigParam param, Object value, int index,
    String path, List<TagStyleValidationIssue> issues) {
  switch (param.type) {
    case 'enum':
      if (value is! String || !(param.options ?? const []).contains(value)) {
        issues.add(TagStyleValidationIssue(
            index: index,
            path: path,
            level: 'error',
            message:
                'valor "$value" inválido (admitidos: ${(param.options ?? const []).join(', ')})'));
      }
      break;
    case 'number':
      if (value is! num) {
        issues.add(TagStyleValidationIssue(
            index: index, path: path, level: 'error', message: 'valor "$value" no es un número'));
        break;
      }
      if (param.min != null && value < param.min!) {
        issues.add(TagStyleValidationIssue(
            index: index, path: path, level: 'error', message: '$value < mínimo ${param.min}'));
      }
      if (param.max != null && value > param.max!) {
        issues.add(TagStyleValidationIssue(
            index: index, path: path, level: 'error', message: '$value > máximo ${param.max}'));
      }
      break;
    case 'string':
      if (value is! String) {
        issues.add(TagStyleValidationIssue(
            index: index, path: path, level: 'error', message: 'valor "$value" no es un string'));
      }
      break;
  }
}

/// Valida un [TagStyle] suelto. `true` si no tiene issues `error`.
bool isTagStyleValid(TagStyle ts) {
  final issues = <TagStyleValidationIssue>[];
  _collectTagStyleIssues(ts, -1, 'tagStyle', issues);
  return !issues.any((i) => i.level == 'error');
}

/// Valida el array completo (`albumSchema.tagStyles`). Además avisa (`warn`) de
/// valores de tag duplicados (dos estilos para el mismo `value` → ambigüedad).
TagStyleValidationResult validateTagStyles(List<TagStyle> tagStyles) {
  final issues = <TagStyleValidationIssue>[];

  for (var index = 0; index < tagStyles.length; index++) {
    _collectTagStyleIssues(tagStyles[index], index, 'tagStyles[$index]', issues);
  }

  // Duplicados por `value` → warn (se aplicaría el primero).
  final seen = <String, int>{};
  for (var index = 0; index < tagStyles.length; index++) {
    final ts = tagStyles[index];
    if (ts.value.trim().isEmpty) continue;
    final prev = seen[ts.value];
    if (prev != null) {
      issues.add(TagStyleValidationIssue(
          index: index,
          path: 'tagStyles[$index].value',
          level: 'warn',
          message:
              'valor de tag "${ts.value}" duplicado (ya definido en tagStyles[$prev]); se aplicaría el primero'));
    } else {
      seen[ts.value] = index;
    }
  }

  return TagStyleValidationResult(
      valid: !issues.any((i) => i.level == 'error'), issues: issues);
}

File: packages/core/src/tag-styles.ts
/**
 * `tag-styles.ts` — KRO-30. Validador puro de `TagStyle[]` contra el catálogo
 * de efectos visuales (`visual-effects.ts`).
 *
 * Dado uno o varios `TagStyle`, comprueba que:
 *  - `value` es un string no vacío.
 *  - `effect` existe en el registry de efectos visuales.
 *  - cada key de `config` pertenece a los params declarados por ese efecto.
 *  - cada valor de `config` cae dentro del espacio válido del param
 *    (enum → en `options`; number → dentro de `min`/`max`; string → string).
 *  - los params `optional: false` están presentes.
 *
 * Útil para:
 *  - Studio: el mini-editor de estilos de tag no deja guardar efectos/config inválidos.
 *  - Backend: validar antes de persistir `albumSchema.tagStyles`.
 *  - Flutter (KRO-65/83): ignorar/avisar tagStyles que no puede renderizar.
 *
 * Pure: sin side effects, sin I/O. Mismas entradas → mismas issues.
 */

import type { TagStyle } from './types';
import { getVisualEffect } from './registries/visual-effects';

// ─────────────────────────────────────────────────────────────────────────
// Tipos públicos
// ─────────────────────────────────────────────────────────────────────────

export interface TagStyleValidationIssue {
  /** Índice del TagStyle en el array validado (-1 si se validó uno suelto). */
  index:   number;
  /** Path JSON-style al campo que falló (e.g. `tagStyles[0].config.intensity`). */
  path:    string;
  /** Mensaje legible en español. */
  message: string;
  /** Severity: error = no aplicable; warn = aplicable con fallback. */
  level:   'error' | 'warn';
}

export interface TagStyleValidationResult {
  /** True si no hay issues `error`. Warnings no invalidan. */
  valid:  boolean;
  /** Todos los issues encontrados (errors + warnings). */
  issues: TagStyleValidationIssue[];
}

// ─────────────────────────────────────────────────────────────────────────
// API pública
// ─────────────────────────────────────────────────────────────────────────

/**
 * Valida UN `TagStyle` y empuja issues a `issues` (con el `index` y `prefix`
 * dados). Reutilizado por `validateTagStyles` y por `isTagStyleValid`.
 */
function collectTagStyleIssues(
  ts:     TagStyle,
  index:  number,
  prefix: string,
  issues: TagStyleValidationIssue[],
): void {
  // 1. value no vacío.
  if (typeof ts.value !== 'string' || ts.value.trim() === '') {
    issues.push({ index, path: `${prefix}.value`, level: 'error', message: 'el valor de la tag no puede estar vacío' });
  }

  // KRO-122 — Foil PERSONALIZADO: no es un efecto de catálogo, sino capas
  // propias. Se valida que haya al menos una capa y que cada una tenga textura.
  if (ts.effect === 'custom_foil' || (ts.customLayers && ts.customLayers.length > 0)) {
    const layers = ts.customLayers ?? [];
    if (layers.length === 0) {
      issues.push({ index, path: `${prefix}.customLayers`, level: 'error', message: 'el foil personalizado necesita al menos una capa' });
    }
    layers.forEach((l, li) => {
      if (typeof l.textureUrl !== 'string' || l.textureUrl.trim() === '') {
        issues.push({ index, path: `${prefix}.customLayers[${li}].textureUrl`, level: 'error', message: 'la capa de foil necesita una textura (URL)' });
      }
    });
    return;
  }

  // 2. effect existe en el catálogo.
  const effect = getVisualEffect(ts.effect);
  if (!effect) {
    issues.push({ index, path: `${prefix}.effect`, level: 'error', message: `el efecto "${ts.effect}" no existe en el catálogo de efectos visuales` });
    return; // sin definición del efecto no podemos validar config.
  }

  const params   = effect.config;
  const paramById = new Map(params.map(p => [p.key, p]));
  const config    = ts.config ?? {};

  // 3. cada key de config pertenece al efecto.
  for (const key of Object.keys(config)) {
    const param = paramById.get(key);
    if (!param) {
      issues.push({ index, path: `${prefix}.config.${key}`, level: 'error', message: `"${key}" no es una opción de config del efecto "${ts.effect}"` });
      continue;
    }
    const value = config[key];
    validateConfigValue(param, value, index, `${prefix}.config.${key}`, issues);
  }

  // 4. params requeridos presentes.
  for (const param of params) {
    if (param.optional === false && !(param.key in config)) {
      issues.push({ index, path: `${prefix}.config.${param.key}`, level: 'error', message: `el efecto "${ts.effect}" requiere config "${param.key}"` });
    }
  }
}

function validateConfigValue(
  param:  { key: string; type: 'enum' | 'number' | 'string'; options?: string[]; min?: number; max?: number },
  value:  unknown,
  index:  number,
  path:   string,
  issues: TagStyleValidationIssue[],
): void {
  switch (param.type) {
    case 'enum': {
      if (typeof value !== 'string' || !(param.options ?? []).includes(value)) {
        issues.push({ index, path, level: 'error', message: `valor "${String(value)}" inválido (admitidos: ${(param.options ?? []).join(', ')})` });
      }
      break;
    }
    case 'number': {
      if (typeof value !== 'number' || Number.isNaN(value)) {
        issues.push({ index, path, level: 'error', message: `valor "${String(value)}" no es un número` });
        break;
      }
      if (param.min !== undefined && value < param.min) {
        issues.push({ index, path, level: 'error', message: `${value} < mínimo ${param.min}` });
      }
      if (param.max !== undefined && value > param.max) {
        issues.push({ index, path, level: 'error', message: `${value} > máximo ${param.max}` });
      }
      break;
    }
    case 'string': {
      if (typeof value !== 'string') {
        issues.push({ index, path, level: 'error', message: `valor "${String(value)}" no es un string` });
      }
      break;
    }
  }
}

/**
 * Valida un `TagStyle` suelto. `true` si no tiene issues `error`.
 */
export function isTagStyleValid(ts: TagStyle): boolean {
  const issues: TagStyleValidationIssue[] = [];
  collectTagStyleIssues(ts, -1, 'tagStyle', issues);
  return !issues.some(i => i.level === 'error');
}

/**
 * Valida un array de `TagStyle` (el `albumSchema.tagStyles` completo).
 * Además de validar cada entry, avisa (`warn`) de valores de tag duplicados
 * (dos estilos para el mismo `value` → ambigüedad de cuál aplicar).
 */
export function validateTagStyles(tagStyles: TagStyle[]): TagStyleValidationResult {
  const issues: TagStyleValidationIssue[] = [];

  tagStyles.forEach((ts, index) => {
    collectTagStyleIssues(ts, index, `tagStyles[${index}]`, issues);
  });

  // Duplicados por `value` → warn (no bloquea, pero el render elegiría el primero).
  const seen = new Map<string, number>();
  tagStyles.forEach((ts, index) => {
    if (typeof ts.value !== 'string' || ts.value.trim() === '') return;
    const prev = seen.get(ts.value);
    if (prev !== undefined) {
      issues.push({ index, path: `tagStyles[${index}].value`, level: 'warn', message: `valor de tag "${ts.value}" duplicado (ya definido en tagStyles[${prev}]); se aplicaría el primero` });
    } else {
      seen.set(ts.value, index);
    }
  });

  return { valid: !issues.some(i => i.level === 'error'), issues };
}
